src/LanguageServer/Protocol/Handler/DocumentChanges/DidChangeHandler.cs

```diff
@@ -28,15 +30,52 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(DidChangeTextDocumentPar
     {
         var text = context.GetTrackedDocumentSourceText(request.TextDocument.Uri);
 
+        text = GetUpdatedSourceText(request.ContentChanges, text);
+
+        context.UpdateTrackedDocument(request.TextDocument.Uri, text);
+
+        return SpecializedTasks.Default<object>();
+    }
+
+    private static SourceText GetUpdatedSourceText(TextDocumentContentChangeEvent[] contentChanges, SourceText text)
+    {
+        var areChangesOrdered = true;
+
         // Per the LSP spec, each text change builds upon the previous, so we don't need to translate any text
-        // positions between changes, which makes this quite easy. See
+        // positions between changes. See
         // https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#didChangeTextDocumentParams
         // for more details.
-        foreach (var change in request.ContentChanges)
-            text = text.WithChanges(ProtocolConversions.ContentChangeEventToTextChange(change, text));
+        //
+        // If the host sends us changes in a way such that no earlier change can affect the position of a later change,
+        // then we can merge the changes into a single TextChange, allowing creation of only a single new
+        // source text.
+        for (var i = 1; i < contentChanges.Length; i++)
+        {
+            var prevChange = contentChanges[i - 1];
+            var curChange = contentChanges[i];
 
-        context.UpdateTrackedDocument(request.TextDocument.Uri, text);
+            if (prevChange.Range.Start.CompareTo(curChange.Range.End) < 0)
```

Comment:
Its early and I'm not 100% sure of this.  Is it possible we need to compare against all previous changes?  For example we have two disparate changes, then one afterwards that overlaps just the first one?

Reply:
Chatted in person, let me know if you still have concerns.